### PR TITLE
ci(framework:skip) Add e2e Docker Compose directory for tests

### PR DESCRIPTION
**Pull request by chongshenng**
### Description

There are no basic templates to run Flower (Next) end-to-end with Docker. 

### Related issues/PRs
N/A

## Proposal
Add a self-contained directory that executes FL training using Flower `ClientApp` and `ServerApp` end-to-end. Use Docker Compose v2.

### Explanation
This is the starting point to add a CI test with `testcontainers`.

### Checklist

- [X] Implement proposed change
- ~~[ ] Write tests~~
- ~~[ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)~~
- [X] Update the changelog entry below
- [X] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Changelog entry

<skip>

### CODE REVIEW

File: e2e/docker/compose.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+services:
+  # create a SuperLink service
+  superlink:
+    image: docker.io/flwr/superlink:1.9.0.dev20240513
+    ports:
+      - "9091:9091"
+      - "9092:9092"
+    networks:
+      - flwr-net
+    hostname: flwr-superlink
```
Comment (Robert-Steiner) [inline e2e/docker/compose.yaml:12]: When using the Docker-Compose the default network by default uses the service name (`superlink`) as hostname. Therefore we can remove the `networks` config in each service and change the SuperLink URL to `"superlink:9092”` or `"superlink:9091”`. 
```suggestion
```

**Review by Robert-Steiner — COMMENTED** [review]
Left some minor suggestions. Overall looks good!

File: e2e/docker/compose.yaml (hunk 1 above)
Comment (chongshenng) [reply e2e/docker/compose.yaml:12]: Ah, cool! Added.

File: e2e/docker/compose.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+name: flwr-containers-e2e-test
+
+services:
+  # create a SuperLink service
+  superlink:
+    image: docker.io/flwr/superlink:1.9.0.dev20240513
```
Comment (jafermarq) [inline e2e/docker/compose.yaml:6]:
```suggestion
    image: docker.io/flwr/superlink:nightly
```

Comment (jafermarq) [reply e2e/docker/compose.yaml:6]: I wonder if we should better point it to the latest nightly. If so, let's change it in the `ServerApp` and `ClientApp` Dockerfile too.

Comment (chongshenng) [reply e2e/docker/compose.yaml:6]: That's a good point. I'll update it.

File: e2e/docker/README.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Flower containers testing
+
+This directory is used to test Flower containers in a minimum scenario, that is, with 2 clients and without HTTPS. The FL setup uses PyTorch, the CIFAR10 dataset, and a CNN. This is mainly to test the functionalities of Flower in a containerized architecture. It can be easily extended to test more complex communication set-ups.
+
+It uses a subset of size 1000 for the training data and 10 data points for the testing.
+
+To execute locally, run the following in CLI:
+``` shell
+$ docker compose up -d --remove-orphans --force-recreate --build
```
Comment (Robert-Steiner) [inline e2e/docker/README.md:9]: If we want to make sure we are always using the latest `nightly` version, we need to run the following:

```
# pulls the latest supernode and serverapp nightly
docker compose build --pull 
# pulls the latest superlink nightly
docker compose up -d --remove-orphans --force-recreate --pull always 
```

File: e2e/docker/supernode.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM flwr/supernode:nightly
+
+WORKDIR /app
+RUN python -m pip install -U --no-cache-dir\
+    "flwr-datasets["vision"]>=0.1.0,<1.0.0" \
+    torch==2.2.1 \
+    torchvision==0.17.1 \
+    && pyenv rehash
```
Comment (tanertopal) [inline e2e/docker/supernode.Dockerfile:8]: Wouldn't it be better to have a `pyproject.toml` and then just copy that? I usually like that pattern more then tracking dependencies in a Dockerfile. Not sure though. I might be missing something.


```suggestion
COPY pyproject.toml ./
RUN python -m pip install -U --no-cache-dir . && pyenv rehash
```

Comment (chongshenng) [reply e2e/docker/supernode.Dockerfile:8]: @tanertopal That's a valid point. I don't have a strong argument for using this apart from possibly running `pytest` against different versions of `torch` and `torchvision`. Will add `pyproject.toml` to it.

File: e2e/docker/README.md (hunk 3 above)
Comment (chongshenng) [reply e2e/docker/README.md:9]: Nice! Will update it.

File: e2e/docker/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "quickstart-pytorch"
```
Comment (tanertopal) [inline e2e/docker/pyproject.toml:6]:
```suggestion
name = "e2e-docker"
```

File: e2e/docker/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+version = "0.1.0"
+description = "PyTorch Federated Learning Quickstart with Flower"
```
Comment (tanertopal) [inline e2e/docker/pyproject.toml:8]:
```suggestion
name = "e2e-docker"
version = "0.1.0"
description = "TOML used to define dependencies in a E2E test"
```

**Review by tanertopal — APPROVED** [review]
lgtm

**Review by tanertopal — APPROVED** [review]

**Pull request merged**